overlord/snapstate/check_snap.go

```diff
@@ -369,8 +369,43 @@ func checkBases(st *state.State, snapInfo, curInfo *snap.Info, flags Flags) erro
 	return fmt.Errorf("cannot find required base %q", snapInfo.Base)
 }
 
+func checkEpochs(_ *state.State, snapInfo, curInfo *snap.Info, _ Flags) error {
+	if curInfo == nil {
+		return nil
+	}
+	if snapInfo.Epoch.CanRead(curInfo.Epoch) {
+		return nil
+	}
+	desc := "local snap"
+	if snapInfo.SideInfo.Revision.Store() {
+		desc = fmt.Sprintf("new revision %s", snapInfo.SideInfo.Revision)
+	}
+
+	return fmt.Errorf("cannot refresh %q to %s with epoch %s, because it can't read the current epoch of %s", snapInfo.InstanceName(), desc, snapInfo.Epoch, curInfo.Epoch)
```

Comment:
These mostly make sense to me. My only general comment on these error messages is that it isn't clear what *it* is referring to in `because it can't read the current epoch`. Should it be *the current snap instance*? Also, should the word *read* be made less ambiguous? Does the current epoch not exist, or is it inaccessible? Can we list available epochs for the current instance?

Reply:
The epoch of a snap revision is a single thing, and it looks like, for example, `0`, or `2*`, or in the worst case something like `{"read":[1,2,3,4],"write":[2,3,4]}`. It's a note from the snap developer saying "this snap reads and writes data in these formats" (where what a format is is intentionally vague), so if you see a snap with epoch`2*` it's because the snap developer thinks it can read formats 1 and 2, and writes format 2. More (perhaps better) docs here: https://forum.snapcraft.io/t/epochs-stepped-upgrades/1757
The store does not give a client a snap with an epoch it can't use, so this error would only be seen by a user installing a snap locally, or doing `snap try` (or bugs).
The more typical case is a developer building a snap and installing it for testing.

In any case, with that context, the error message the user would see would be something like

    $ snap install --dangerous ./test-snapd-epoch_2_all.snap 
    error: cannot install snap file: cannot refresh "test-snapd-epoch" to local snap with epoch 2,
           because it can't read the current epoch of 1

(if we were in the supposed-to-be-impossible case of the store giving us something we can't use, the "local snap" would be "new revision NNN").

So to come back to your original point, the "it" in that sentence is a bit ambiguous because it could (AIUI) refer to either the "local snap" or to "epoch 2", but those two are the same thing really so I don't think any meaning is lost.

Reply:
thanks for this great overview @chipaca - that makes complete sense, and I think you're right about keeping it the same.